Add SingletonSymIntNode

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107089
* #107839

Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic.

soulitzer committed:
[WIP] Add SingletonSymIntNode

[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

ezyang commented on c10/core/Scalar.h:
```diff
@@ -300,12 +300,13 @@ class C10_API Scalar {
 
   Scalar(c10::SymBool sb) {
-    if (sb.is_symbolic()) {
+    // Why do we store bool in a double? (preexisting)
```
Seems like a bug here. When we look at other code I see that bool is accessed via int field

ezyang commented on c10/core/SingletonSymNodeImpl.h:
```diff
@@ -0,0 +1,121 @@
+
+namespace c10 {
+
+// An object that only defines the equality operator.
```
With itself!

A little more motivating explanation here is always helpful.

soulitzer replied:
Hmm but we do allow `singleton(1) == singleton(1)` and that seems more like normal equality to me?

We only get the "with itself" idea once we have something like `get_next_unique_singleton(offset) != get_next_unique_singleton(offset.clone())`.

ezyang commented on c10/core/SingletonSymNodeImpl.h:
```diff
@@ -0,0 +1,121 @@
+// An object that only defines the equality operator.
+class C10_API SingletonSymNodeImpl : public SymNodeImpl {
+ public:
+  SingletonSymNodeImpl(int64_t val) : val_(val) {}
```
Gotta be very careful with this constructor, people might think this denotes regular int semantics

soulitzer replied:
Also related https://github.com/pytorch/pytorch/pull/107089#discussion_r1294006232? 
Would it be sufficient to have a comment saying that no one should call this constructor directly?

soulitzer replied:
Made the construct explicit and added a comment telling folks to not use this directly.

ezyang commented on c10/core/SingletonSymNodeImpl.h:
```diff
@@ -0,0 +1,121 @@
+
+  int64_t guard_int(const char* file, int64_t line) override {
+    // TODO: when is this used?
+    TORCH_CHECK(false);
```
This would be called if you did, say, int(nt.size(1)). Which is not satisfiable since these singletons are not representable as plain ints

ezyang commented on c10/core/SingletonSymNodeImpl.h:
```diff
@@ -0,0 +1,121 @@
+    TORCH_CHECK(
+        c,
+        "SingletonSymNode can only be compared with SingletonSymNode, but got ",
+        other->str());
```
It's not obvious to me you should error here, another reasonable answer would be to return false

soulitzer replied:
That's a good point. I haven't thought about this choice deeply, though maybe its ok to potentially relax later as a follow up?

ezyang commented on c10/core/SingletonSymNodeImpl.h:
```diff
@@ -0,0 +1,121 @@
+  }
+
+  c10::SymNode eq(const c10::SymNode& other) override {
+    c10::optional<int64_t> c = other->singleton_int();
```
You are going to need some multi dispatch here eventually. At some point we will want to create symbolic ints that range over singletons and so you will need to be able to answer questions like s0 == j3

ezyang commented on c10/core/SingletonSymNodeImpl.h:
```diff
@@ -0,0 +1,121 @@
+#undef DEFINE_NOT_SUPPORTED
+
+ private:
+  int64_t val_;
```
I assume these are allocated as uniques but where is the next allocatable int stored?

soulitzer replied:
Unless there's some place we need to create them in cpp (I may need to think deeper about that), we can do something like the following in Python:
```python
_counter = 0
_map = WeakTensorKeyDictionary()

def get_tensor_handle_int(tensor):
    global _counter
    if tensor not in _map:
        _map[tensor] = _counter
        _counter += 1
    return torch._C.get_singleton_int(_map[tensor])

```

ezyang commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -738,6 +739,11 @@ void initDispatchBindings(PyObject* module) {
+
+  m.def("get_singleton_int", [](int64_t data) {
+    return c10::SymInt(
+        c10::SymNode(c10::make_intrusive<c10::SingletonSymNodeImpl>(data)));
```
Hmm, I'm not sure I want to let people generate singletons by number, it may be safer to offer only a fresh singleton api

soulitzer replied:
See other comment https://github.com/pytorch/pytorch/pull/107089#discussion_r1294006232

soulitzer replied:
Made this a private API.

soulitzer commented on c10/core/SymInt.cpp:
```diff
@@ -39,14 +39,18 @@ bool SymInt::has_hint() const {
       } else {                                                       \
         auto b = sci.toSymNode();                                    \
-        return RET(b->wrap_int(*ma)->METHOD(b));                     \
+        return RET(b->METHOD(b->wrap_int(*ma)));                     \
```
This is the reason for most (if not all?) failures in CI right now. (It changes assymetric ops like gt to lt and vice versa, whoops!).

I'm still thinking through different ways to do this, but properly doing multiple dispatch could be one way...

soulitzer replied:
I did the simplest thing possible which was to override eq in ConstantSymNode to call into SingletonSymNode. This doesn't do too much in this version, just gives us a slightly nicer error instead of the default "NYI".

soulitzer committed:
Update on "[WIP] Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer committed:
Update on "Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

ezyang commented on c10/test/core/SymInt_test.cpp:
```diff
@@ -21,4 +22,32 @@ TEST(SymIntTest, ConcreteInts) {
+  ASSERT_TRUE(a != c);
+
+  // Tentaively throw an error when comparing with a non-singleton, this is not
+  // necessarily the right behavior.
```
Yeah, I could see an argument that comparison with integers just returns False. We can change this when we get more experience with it in the frontend.

ezyang commented on c10/core/SymNodeImpl.h:
```diff
@@ -164,16 +164,22 @@ class C10_API SymNodeImpl : public c10::intrusive_ptr_target {
+    return c10::nullopt;
+  }
+  virtual c10::optional<int64_t> constant_int() {
+    return c10::nullopt;
```
This feels like XLA might need to implement it

soulitzer replied:
Oh wow didn't realize they have XLASymNodeImpl
Hmmm they didn't implement large_negative_number either though

ezyang commented on c10/core/SymInt.h:
```diff
@@ -209,15 +209,10 @@ class C10_API SymInt {
-      return d;
-    }
-    return c10::nullopt;
+    return node->maybe_as_int();
```
This is something of a preexisting problem, but I wonder why I didn't just have `maybe_as_int` take care of everything...

soulitzer replied:
Filing an issue to do this later https://github.com/pytorch/pytorch/issues/107873

ezyang commented on c10/core/ConstantSymNodeImpl.h:
```diff
@@ -0,0 +1,92 @@
+  // Currently even if we have this method, we still raise an error when we get
+  // to SingletonSymNode::eq since comparing with non-singleton is disallowed.
+  // However, we may change that behavior in the future.
+  c10::SymNode eq(const c10::SymNode& other) override {
```
nit: put this in the cpp file, instead of the header, better for binary size and easier to change later.

ezyang commented on c10/core/ConstantSymNodeImpl.h:
```diff
@@ -0,0 +1,92 @@
+  c10::SymNode eq(const c10::SymNode& other) override {
+    TORCH_INTERNAL_ASSERT(other->singleton_int().has_value());
+    return other->eq(c10::SymNode(make_intrusive<ConstantSymNodeImpl<int64_t>>(
+        c10::get<int64_t>(value_))));
```
Why can't you just pass `*this` here?

soulitzer replied:
Because it's expecting a SymNode (intrusive ptr to SymNodeImpl) but `*this` gives us ConstantSymNodeImpl is a subclass of SymNodeImpl. 

Updated to just do intrusive_from_this instead:
```cpp
c10::raw::intrusive_ptr::incref(this);
return other->eq(c10::intrusive_ptr<ConstantSymNodeImpl<T>>::reclaim(this));
```

ezyang commented on c10/core/ConstantSymNodeImpl.h:
```diff
@@ -0,0 +1,92 @@
+  // Temporary hack to avoid having to implement multiple dispatch for now
+  // Currently even if we have this method, we still raise an error when we get
+  // to SingletonSymNode::eq since comparing with non-singleton is disallowed.
+  // However, we may change that behavior in the future.
```
Can we do basic reflexivity? ConstantSymNode should always compare equal with itself, if the val is the same.

soulitzer replied:
Reflexivity already happens - If both sides are ConstantSymNode we would've unwrapped in SymInt/SymBool already. We assert that we only reach here in the case where `plain.op(singleton)` and plain gets promoted to constant, so that becomes `constant.op(singleton)`.

soulitzer commented on c10/core/SingletonSymNodeImpl.cpp:
```diff
@@ -0,0 +1,3 @@
+#include <c10/core/SingletonSymNodeImpl.h>
+
+namespace c10 {} // namespace c10
```
Yeah ik, unfortunately this is necessary to get Windows to build.

ezyang commented on c10/core/SymBool.cpp:
```diff
@@ -1,84 +1,79 @@
 
+SymNode SymBool::promoteToConstantSymNode() const {
+  TORCH_CHECK(!is_heap_allocated())
+  return SymNode(c10::make_intrusive<ConstantSymNodeImpl<bool>>(data_));
```
I don't really understand why you need this. Can you tell me more?

soulitzer replied:
Good point, we actually we don't. I can just do .expect_bool() instead

ezyang reviewed (approved):
okey dokey

soulitzer committed:
Update on "Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer committed:
Update on "Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer committed:
Update on "Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer committed:
Update on "Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer committed:
Update on "Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer committed:
Update on "Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer committed:
Update on "Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer committed:
Update on "Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer committed:
Update on "Add SingletonSymIntNode"


Adds `SingletonSymNodeImpl` (alternatively, `SkolemSymNodeImpl`). This is a int-like object that only allows  the`eq` operation; any other operation produces an error.

The main complexity is that we require operations that dispatch to SymNode must take and return SymNodes, but when performing operations involving `SingletonSymNodeImpl`, operations involving SymNode can return non-SymNode bools.  For more discussion see [here](https://docs.google.com/document/d/18iqMdnHlUnvoTz4BveBbyWFi_tCRmFoqMFdBHKmCm_k/edit)
- Introduce `ConstantSymNodeImpl` a generalization of `LargeNegativeIntSymNodeImpl` and replace usage of `LargeNegativeIntSymNodeImpl`  in SymInt.
- Also use ConstantSymNodeImpl to enable SymBool to store its data on a SymNode. Remove the  assumption that if SymBool holds a non-null SymNode, it must be symbolic. 





[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge
